### DISCUSSION TITLE
Modbus RTU master feature

### DISCUSSION THREAD

**Discussion by taliesin** [dzurikmiroslav/esp32-evse#68] [Ideas]
Currently my wallbox incorporates a 3phase meter (Eastron SDM630) which is connected to SmartEVSE V3 via RS485. This project does support Modbus RTU, but only as a slave.
I'll happily implement the actual communication part with the meter in Lua, but this will at least need a driver to access the serial line and a configuration to enable master or slave mode.

ESP-IDF already provides a component for both master and slave mode, wouldn't it make sense to use that in the project?

**Comment by dzurikmiroslav** [comment]
Yes, ESP-IDF has a Modbus library, but it was not suitable (reading register values ​​was from a pointer, the getter function could not be called). Implementing a Modbus serial/TCP slave was not very difficult...

I have just started working on unit tests after a large number of pull request :)

I can focus on extending the Lua API if the pull requests will wait?

**Comment by taliesin** [comment]
Yeah, I just walked through the ESP modbus API, it's really a bit awkward and your implementation is very straight-forward.
Keep going with the unit tests, it's definitely necessary.

I'm currently reviewing the serial communication stuff, so probably the pull requests will continue :-)

May I ask a side questions here? The board_config states pins and serials individually, does anything speak against arrays there?
I think it would make init and access quite a bit simpler.

**Comment by taliesin** [comment]
I've just completed the rework of board_config.c and serial.c using arrays for the aux stuff and the serials.
This removed about 130 lines of code and reduced the code size (-Os) by about 1.6kB and I think complexity went down also.

If you are working on unit tests concerning these blocks, it might make sense to incorporate the pull requests first. This last one will only go out, when you have merged the current one.

I'll be happy to assist you writing tests, given the framework you chose.

**Reply by dzurikmiroslav** [reply]
Good, was it necessary to change the names of the variables in `board.cfg` file?

I just push branch [unity-test](https://github.com/dzurikmiroslav/esp32-evse/tree/unity-test) there is folder `unity-app`. 
It is a unit test as used in ESP-IDF, I can only test the evse component functionality, because peripherals are mocked...

It runs directly on ESP32 board (it could also run on Wokwi), I would be interested in your opinion. How to test other components, for each component make separate test binary?

Do you know a better framework for testing?

**Comment by taliesin** [comment]
> Good, was it necessary to change the names of the variables in board.cfg file?

No, everything stays compatible.

Concerning the test framework, I've been working with google test and boost test, both are target independent, but I think unity should be OK. I usually separate my projects in platform specific parts and non-platform code, this way you can test most of the code on the host, but the separation can be a hassle.

We can try to build parts of the application for Linux, but this will result in quite some effort. Let's go and think later (most of the time the best solution for test code, at least in my opinion).

I need to look into unity and will try to provide test code for the changes I want you to merge, let's see if it works out. Probably we should open a new discussion concerning testing.
